### PR TITLE
Add differential steering

### LINKED ISSUE #17: Add differential steering
Add:
- [ ] Keyboard control of the robot in simulation
- [x] Differential controller that takes the geometry_msgs/Twist message to control the robot. Inspiration can be taken from the diff_drive_controller package.

### PR DESCRIPTION
**Pull request by SmejoTT**
Closes #17

### CODE REVIEW

File: lib/steering/src/diffSteering.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+void shutDownMotors() {
+  for (int i = 0; i < std::size(wheels); i++) {
+    wheels[i]->setVelocity(0.0);
+  }
+}
+
+void keyboardInput() {
+  double speed = 2.0;
+  webots::Keyborad keyboard;
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:60]:
```suggestion
  webots::Keyboard keyboard;
```

File: lib/steering/src/diffSteering.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        wheels[1]->setVelocity(-speed);
+        wheels[3]->setVelocity(-speed);
+        break;
+      case keyboard.LEFT:
+        wheels[0]->setVelocity(-speed);
+        wheels[2]->setVelocity(-speed);
+        wheels[1]->setVelocity(speed);
+        wheels[3]->setVelocity(speed);
+        break;
+}
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:88]:
```suggestion
}

```

File: src/full_controller.cpp [hunk 3]
```diff
@@ -30,12 +31,24 @@ int main(int argc, char **argv) {
   webots::Lidar *wb_lidar = robot->getLidar(lidar_name);
   webots::GPS *wb_gps = robot->getGPS(gps_name);
   webots::InertialUnit *wb_imu = robot->getInertialUnit(imu_name);
+  webots::Motor *front_left_wheel = robot->getMotor("front_left_wheel");
```
Comment (Siliconlad) [inline src/full_controller.cpp:34]: Can you not combine this into:
```cpp
webots::Motor *motors[4] = {
    robot->getMotor("front_left_wheel"),
    robot->getMotor("front_right_wheel"),
    robot->getMotor("rear_left_wheel"),
    robot->getMotor("front_right_wheel")
};
```
Or if you don't like that:
```cpp
webots::Motor *motors[4];
motors[0] = robot->getMotor("front_left_wheel");
motors[1] = robot->getMotor("front_right_wheel");
motors[2] = robot->getMotor("rear_left_wheel");
motors[3] = robot->getMotor("front_right_wheel");
};
```

File: lib/steering/include/diffSteering.hpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  const double wheel_separation;
+  const double wheel_radius;
+  
+  // Create Subscriber
+  ros::Subscriber cmd_vel_sub;
+
+  void subscribeVelocity(const geometry_msgs::Twist& cmd);
+  void turnOnMotors();
+  void shutDownMotors();
+  void keyboardInput();
```
Comment (Siliconlad) [inline lib/steering/include/diffSteering.hpp:24]: You'll need to move this method (`keyboardInput`) into the public section of this class declaration if you want to call it from outside the class like you do in `full_controller_keyboard`.

File: src/full_controller_keyboard.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  nh.param("step_size", step_size, 32);
+
+  // Set up Webots
+  webots::Robot *robot = new webots::Robot();
+
+  // Get webot devices
+  webots::Lidar *wb_lidar = robot->getLidar(lidar_name);
+  webots::GPS *wb_gps = robot->getGPS(gps_name);
+  webots::InertialUnit *wb_imu = robot->getInertialUnit(imu_name);
+  webots::Motor *front_left_wheel = robot->getMotor("front_left_wheel");
```
Comment (Siliconlad) [inline src/full_controller_keyboard.cpp:34]: See above for making these more concise.

**Review by Siliconlad — CHANGES_REQUESTED** [review]
Your `keyboardInput()` method could probably be its own ROS node in autonomous? It can listen to keyboard input and just publish `geometry_msgs/Twist` messages. That way we don't need different controllers.

File: CMakeLists.txt [hunk 6]
```diff
@@ -48,18 +48,22 @@ set (
...
 ## Declare a C++ library
 add_library(sensors lib/sensors/src/lidar.cpp lib/sensors/src/gps.cpp lib/sensors/src/inertialUnit.cpp)
+add_library(steering lib/steering/src/diffSteering.cpp)
 target_link_libraries(sensors ${WEBOTS_LIB} ${catkin_LIBRARIES})
+target_link_libraries(steering ${WEBOTS_LIB} ${catkin_LIBRARIES})
 
 ## Declare a C++ executable
 add_executable(full_controller src/full_controller.cpp)
 target_link_libraries(full_controller sensors ${WEBOTS_LIB} ${catkin_LIBRARIES})
+target_link_libraries(full_controller steering ${WEBOTS_LIB} ${catkin_LIBRARIES})
```
Comment (Siliconlad) [inline CMakeLists.txt:66]:
```suggestion
target_link_libraries(full_controller sensors steering ${WEBOTS_LIB} ${catkin_LIBRARIES})
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -48,18 +48,22 @@ set (
 include_directories(
     include
     lib/sensors/include
+    lib/steering/include
     $ENV{WEBOTS_HOME}/include/controller/c
     $ENV{WEBOTS_HOME}/include/controller/cpp
     ${catkin_INCLUDE_DIRS}
 )
 
 ## Declare a C++ library
```
Comment (Siliconlad) [inline CMakeLists.txt:57]:
```suggestion
## Declare C++ libraries
```

File: lib/steering/include/steering/diffSteering.hpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+namespace AutomatED {
+    
+class DiffSteering {
+public:
+  DiffSteering(webots::Motor *motors[], ros::NodeHandle *ros_handle);
+  ~DiffSteering();
+  void keyboardInput();
+
+private:
+  webots::Motor **wheels;
```
Comment (Siliconlad) [inline lib/steering/include/steering/diffSteering.hpp:14]: Any reason why we use `webots::Motor**` here but `webots::Motor* motors[]` in the constructor?

File: lib/steering/src/diffSteering.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+void DiffSteering::velocityCallback(const geometry_msgs::Twist& cmd){
+  double linear_vel = cmd.linear.x;
+  double angular_vel = cmd.angular.z;
+  
+  const double vel_left  = 
+          (linear_vel - angular_vel * wheel_separation / 2.0)/wheel_radius;
+  const double vel_right =
+          (linear_vel + angular_vel * wheel_separation / 2.0)/wheel_radius;
+
+  std::cout << vel_left;
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:40]: Use `ROS_INFO` or `ROS_DEBUG` instead. See [ROS Logging](https://wiki.ros.org/roscpp/Overview/Logging).

File: lib/steering/src/diffSteering.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  // Set velocity to left wheels
+  wheels[0]->setVelocity(vel_left);
+  wheels[2]->setVelocity(vel_left);
+  // Set velocity to right wheels
+  wheels[1]->setVelocity(vel_right);
+  wheels[3]->setVelocity(vel_right);
+}
+
+void DiffSteering::turnOnMotors() {
+  for (int i = 0; i < 4; i++) {
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:50]: Can you not check for the length of the array in the `for` loop.

File: lib/steering/src/diffSteering.cpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    wheels[i]->setPosition(INFINITY);
+    wheels[i]->setVelocity(0.0);
+  } 
+}
+
+void DiffSteering::shutDownMotors() {
+  for (int i = 0; i < 4; i++) {
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:57]: Same here

**Review by Siliconlad — CHANGES_REQUESTED** [review]
Two other things:
1. I thought we were going to create a separate node for keyboard control? That way we can remove the `keyboardInput` method.
2. Is it possible to include wheel odometry as part of the steering? Or maybe as a separate file inside the steering library? That would be useful for localization.

File: CMakeLists.txt [hunk 12]
```diff
@@ -87,11 +90,20 @@ install(TARGETS sensors
...
     RUNTIME DESTINATION ${CATKIN_GLOBAL_BIN_DESTINATION}
 )
+install(TARGETS steering
+    ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+    LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+    RUNTIME DESTINATION ${CATKIN_GLOBAL_BIN_DESTINATION}
+)
 
 ## Mark cpp header files for installation
 install(DIRECTORY lib/sensors/include/sensors/
```
Comment (Siliconlad) [inline CMakeLists.txt:100]: Actually, I think I got this wrong previously. I think it should instead be:
```suggestion
install(DIRECTORY lib/sensors/include
```

File: CMakeLists.txt [hunk 13]
```diff
@@ -87,11 +90,20 @@ install(TARGETS sensors
...
     DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
 )
+install(DIRECTORY lib/steering/include/steering/
```
Comment (Siliconlad) [inline CMakeLists.txt:103]: Similarly for here:
```suggestion
install(DIRECTORY lib/steering/include
```

File: lib/steering/src/diffSteering.cpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
+#include "ros/ros.h"
+#include "steering/diffSteering.hpp"
+#include "geometry_msgs/Twist.h"
+#include "webots/Motor.hpp"
+#include "webots/Keyboard.hpp"
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:5]: Remove?

File: lib/steering/src/diffSteering.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+  // Set velocity to left wheels
+  wheels[0]->setVelocity(vel_left);
+  wheels[2]->setVelocity(vel_left);
+  // Set velocity to right wheels
+  wheels[1]->setVelocity(vel_right);
+  wheels[3]->setVelocity(vel_right);
+}
+
+void DiffSteering::turnOnMotors() {
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:49]: Do we need this method? Can we not just put it in the constructor where it gets called?

File: lib/steering/src/diffSteering.cpp [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  for (int i = 0; i < wheel_count; i++) {
+    wheels[i]->setPosition(INFINITY);
+    wheels[i]->setVelocity(0.0);
+  } 
+}
+
+void DiffSteering::shutDownMotors() {
```
Comment (Siliconlad) [inline lib/steering/src/diffSteering.cpp:56]: Would `stopMotors` be a better name?

File: worlds/world.wbt [hunk 17]
```diff
@@ -25,7 +25,6 @@ DEF ACKERMAN Robot {
       name "imu"
     }
     GPS {
-      name "gps"
```
Comment (Siliconlad) [inline worlds/world.wbt:28]: You removed this name?

**Review by Siliconlad — CHANGES_REQUESTED** [review]
After these minor changes, I think I'm happy to merge.

File: worlds/world.wbt (hunk 17 above)
Comment (SmejoTT) [reply worlds/world.wbt:28]: It's missing in the main as well.

Comment (Siliconlad) [reply worlds/world.wbt:28]: Oh ok. I think it would be best to add it just to be explicit even if "gps" is the default.

**Review by Siliconlad — APPROVED** [review]
Looks good!

**Pull request merged**